kvserver: segregate replica metrics in the replicateQueue

amygao9 opened the pull request:
This commit adds separate metrics for tracking adding/removing/rebalancing of voting/non-voting replicas. For instance, the add replica count is an aggregation of voter and non-voter addition counts, which are now also separately tracked. Also added a metric for tracking adding and removing decommissioning replicas.

Fixes [#60694](https://github.com/cockroachdb/cockroach/issues/60694)
Release justification: additive change to add metrics for existing functionality

amygao9 committed:
kvserver: fix testing knob cast type

Previously the wrong object was being casted which caused the type
assertion to fail for NodeLivenessKnobs.  Thus the testing knob was never
attached in some test cases.  This patch ensures the correct object is
casted.  This uncovered a bug where a stale range descriptor was returned
in setup fns, so this commit also patches that by having
`checkReplicaCounts` take a pointer to a range decriptor. Lastly, due to
these fixes we found that a race between the `replicateQueue`'s snapshots and
`raftSnapshotQueue's` causes the `TestReplicateQueueDeadNonVoters` test
to timeout. A new issue #76996 was created to harden this test, but the
test will be skipped for the time being.

Release note: None

aayushshah15 reviewed (commented):
This looks mostly good, thanks for pushing through the tedious aspects!

A general comment is that a lot of lines aren't wrapped to 100 chars. If you don't have `crlfmt` set up, its worth doing so. It'll make all this much easier. https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/154206209/Goland+Tips+and+Tricks#Enable-crlfmt-Watcher

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-:-MvUxZRq8VbGjQ2hSebC:b-3fq6j6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amygao9 and @nvanbenschoten)

___
*[pkg/kv/kvserver/allocator.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUrZee8vX3xOYuuSDG:-MvUrZee8vX3xOYuuSDH:b-2e6we2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/allocator.go#L188)):*
> ```Go
> )
> 
> type replicaStatus int
> ```

Let's give this a brief comment with a description of how its used.
___
*[pkg/kv/kvserver/replicate_queue.go, line 246 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUvg8Z4srBycE7Z0Yc:-MvUvg8Z4srBycE7Z0Yd:b-p9e4wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue.go#L246)):*
> ```Go
> }
> 
> func (metrics *ReplicateQueueMetrics) trackAddReplicaCount(targetType targetReplicaType) {
> ```

Let's add brief comments to these methods.
___
*[pkg/kv/kvserver/replicate_queue.go, line 671 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUvoBp6VEL5O1n0Dgp:-MvUvoBp6VEL5O1n0Dgq:b-eiyv7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue.go#L671)):*
> ```Go
> 	removeIdx int,
> 	dryRun bool,
> 	replicaStatus replicaStatus,
> ```

nit: let's keep `dryRun` as the last parameter. So, let's order this up. Your IDE should let you do this without any tedium.
___
*[pkg/kv/kvserver/store.go, line 2971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUwBY09kVpHvOdN_Zi:-MvUwBY09kVpHvOdN_Zj:ba0v0qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/store.go#L2971)):*
> ```Go
> }
> 
> // ReplicateQueueMetrics returns the store's metric struct.
> ```

Let's say something about the fact that it returns the `replicateQueue`'s metrics. 
___
*[pkg/server/server.go, line 414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvAMUFNAECbNEBJpW6t:-MvAMUFNAECbNEBJpW6u:bkxi1h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc03a18e055a942232981020009abce16bc1ab/pkg/server/server.go#L414)):*
> ```Go
> 
> 	nodeLivenessFn := kvserver.MakeStorePoolNodeLivenessFunc(nodeLiveness)
> 	if nodeLivenessKnobs, ok := cfg.TestingKnobs.NodeLiveness.(kvserver.NodeLivenessTestingKnobs); ok &&
> ```

Does this commit pass these `TestReplicateQueue...` tests on its own? I remember that we learned another bug in one of those tests was offsetting this bug and making the test pass. 

If so, let's move some of those changes into this commit so each of these commits pass these tests. Look up `git add -p`, which should let you stage individual chunks separately.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 499 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxJUSBhvaxBnt9C3W:-MvUxJUSBhvaxBnt9C3X:buyqggw) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L499)):*
> ```Go
> 			return true
> 		}, testutils.DefaultSucceedsSoonDuration, 100*time.Millisecond)
> ```

Let's talk about how rebalances update the addition/removal metric per replica, which is why we're asserting on additions here.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 504 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxCB2Adcl3Peelpqm:-MvUxCB2Adcl3Peelpqn:bfjghqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L504)):*
> ```Go
> 		currentDecommRemovals := store.ReplicateQueueMetrics().RemoveDecommissioningNonVoterReplicaCount.Count()
> 		require.GreaterOrEqualf(t, currentAddCount, previousAddCount+2,
> 			"expected replica removals to increase by at least 2")
> ```

additions*
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxFuy3ARwTnduCbQK:-MvUxFuy3ARwTnduCbQL:b-o0ww3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L506)):*
> ```Go
> 			"expected replica removals to increase by at least 2")
> 		require.GreaterOrEqualf(t, currentRemoveCount, previousRemovalCount+2,
> 			"expected replica removals to increase by at least 2")
> ```

total removals*


<!-- Sent from Reviewable.io -->

amygao9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/allocator.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUrZee8vX3xOYuuSDG:-MvtVw7U3qjQOPvfOlA9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/allocator.go#L188)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Let's give this a brief comment with a description of how its used.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue.go, line 246 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUvg8Z4srBycE7Z0Yc:-MvtVwv9025jj7qgC29B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue.go#L246)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Let's add brief comments to these methods.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 2971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUwBY09kVpHvOdN_Zi:-MvtWO8P59z77p6JLl-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/store.go#L2971)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Let's say something about the fact that it returns the `replicateQueue`'s metrics. 
</blockquote></details>

Done.



_Code quote:_
```Go
// ReplicateQueueMetrics returns the store's metric struct.
```
___
*[pkg/server/server.go, line 414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvAMUFNAECbNEBJpW6t:-MvtnNtl7zu9e4-Xdzuz:b563zh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc03a18e055a942232981020009abce16bc1ab/pkg/server/server.go#L414)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Does this commit pass these `TestReplicateQueue...` tests on its own? I remember that we learned another bug in one of those tests was offsetting this bug and making the test pass. 

If so, let's move some of those changes into this commit so each of these commits pass these tests. Look up `git add -p`, which should let you stage individual chunks separately.
</blockquote></details>

Yes, I believe this commit passes the tests on its own. Although I don't quite remember why the tests were passing before when the testing knob wasn't properly attached. The other bug we found was that the `setupFn` was possibly returning a stale scratchRange, which I fixed in the commit after. Should that bug fix be added to the first commit?
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 499 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxJUSBhvaxBnt9C3W:-MvtaZMp4R24DDyPnrZi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L499)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Let's talk about how rebalances update the addition/removal metric per replica, which is why we're asserting on additions here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 504 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxCB2Adcl3Peelpqm:-MvtaYlcATIp2ERU8etn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L504)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

additions*
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvUxFuy3ARwTnduCbQK:-MvtaiuK11e0rwBAvAbR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be3f57fe3a15b82bd71a1f5380ad9cff6a3d5058/pkg/kv/kvserver/replicate_queue_test.go#L506)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

total removals*
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
___
*[pkg/server/server.go, line 414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvAMUFNAECbNEBJpW6t:-MwDLv4k2DW-uXCGkbD4:b7z7vrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc03a18e055a942232981020009abce16bc1ab/pkg/server/server.go#L414)):*
> Yes, I believe this commit passes the tests on its own. Although I don't quite remember why the tests were passing before when the testing knob wasn't properly attached.

I checked the first commit out and the tests are passing because, as you said, the `setupFn` returns a stale range descriptor (that has no non-voters), so we dont actually end up marking any of the nodes as dead. If you apply the following patch to the first commit, you'll see what I mean: 
```
diff --git a/pkg/kv/kvserver/replicate_queue_test.go b/pkg/kv/kvserver/replicate_queue_test.go
index 5bb943a6e6..875931834b 100644
--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -605,6 +605,7 @@ func TestReplicateQueueDeadNonVoters(t *testing.T) {
 		defer tc.Stopper().Stop(ctx)
 
 		beforeNodeIDs := getNonVoterNodeIDs(scratchRange)
+		require.Len(t, beforeNodeIDs, 2)
 		markDead(beforeNodeIDs)
 		require.Eventually(t, func() bool {
 			ok, err := checkReplicaCount(ctx, tc, scratchRange, 1 /* voterCount */, 2 /* nonVoterCount */)

```

Your next commit fixes this by having `checkReplicaCount` take in a pointer to the range descriptor, I'd recommend moving that to the first commit. 


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @amygao9, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue_test.go, line 646 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MwEhG4W4YiACO5lB3t0:-MwEhG4W4YiACO5lB3t1:b2cu5sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf0ae740dc5b4fff311c16de7adb8429063c2f0/pkg/kv/kvserver/replicate_queue_test.go#L646)):*
> ```Go
> 		_, err := tc.ServerConn(0).Exec(
> 			`ALTER RANGE DEFAULT CONFIGURE ZONE USING num_replicas = 3, num_voters = 1`,
> 		)
> ```

I looked at why the test is failing on CI, and its happening because of the following interaction:

When we mark 2 of these nodes as dead, the `replicateQueue`s of the various leaseholder stores rebalance all those nodes' replicas away. Any of these rebalances may fail if the Raft leader is uncollocated with the leaseholder since that opens the possibility of a race between the Raft snapshot queue on Raft leader racing against the leaseholder's replicateQueue sending a snapshot to the newly added learner during the rebalance.  

This is all fine, except that when this race occurs, the `replicateQueue`'s snapshot fails and we roll the learner back, and mark the new store as "throttled" for 5 seconds (which will make the store not a valid rebalance target for the next 5 seconds). https://github.com/cockroachdb/cockroach/blob/52c2df4fb00d943328dc59df9dc210a40a556ab2/pkg/kv/kvserver/store_pool.go#L44-L53

If the raft snapshot queue wins the race enough times (i.e. around 9 times, which will stall rebalances for 45 seconds), the test fails since the call to `forceScanOnAllReplicationQueues` inside `checkReplicaCount` will take longer than 45 seconds.

To work around this, I'd suggest setting that cluster setting to `1ms`. Locally, I couldn't repro the failure when I did that. I'd recommend doing this in the setupFn of both the `Dead` and `Decommissioning` variants of the test. 

```
diff --git a/pkg/kv/kvserver/replicate_queue_test.go b/pkg/kv/kvserver/replicate_queue_test.go
index fdd56450d9..93624a6df5 100644
--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -638,10 +638,16 @@ func TestReplicateQueueDeadNonVoters(t *testing.T) {
 				},
 			},
 		)
+
+		_, err := tc.ServerConn(0).Exec(
+			`SET CLUSTER SETTING server.failed_reservation_timeout='1ms'`,
+		)
+		require.NoError(t, err)
+
 		// Setup a scratch range on a test cluster with 2 non-voters and 1 voter.
 		scratchKey := tc.ScratchRange(t)
 		scratchRange := tc.LookupRangeOrFatal(t, scratchKey)
-		_, err := tc.ServerConn(0).Exec(
+		_, err = tc.ServerConn(0).Exec(
 			`ALTER RANGE DEFAULT CONFIGURE ZONE USING num_replicas = 3, num_voters = 1`,
 		)
 		require.NoError(t, err)

``` 


<!-- Sent from Reviewable.io -->

amygao9 committed:
kvserver: segregate replica metrics in the replicateQueue

This commit adds separate metrics for tracking adding/removing/rebalancing
of voting/non-voting replicas. For instance, the add replica count is an
aggregation of voter and non-voter addition counts, which are now also
separately tracked. Also added a metric for tracking adding and removing
decommissioning replicas.

Fixes #60694
Release justification: additive change to add metrics for existing
functionality
Release note: None

amygao9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/server/server.go, line 414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MvAMUFNAECbNEBJpW6t:-MwcClUNDkyEhfUPqMbR:b-yy0g7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc03a18e055a942232981020009abce16bc1ab/pkg/server/server.go#L414)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

> Yes, I believe this commit passes the tests on its own. Although I don't quite remember why the tests were passing before when the testing knob wasn't properly attached.

I checked the first commit out and the tests are passing because, as you said, the `setupFn` returns a stale range descriptor (that has no non-voters), so we dont actually end up marking any of the nodes as dead. If you apply the following patch to the first commit, you'll see what I mean: 
```
diff --git a/pkg/kv/kvserver/replicate_queue_test.go b/pkg/kv/kvserver/replicate_queue_test.go
index 5bb943a6e6..875931834b 100644
--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -605,6 +605,7 @@ func TestReplicateQueueDeadNonVoters(t *testing.T) {
 		defer tc.Stopper().Stop(ctx)
 
 		beforeNodeIDs := getNonVoterNodeIDs(scratchRange)
+		require.Len(t, beforeNodeIDs, 2)
 		markDead(beforeNodeIDs)
 		require.Eventually(t, func() bool {
 			ok, err := checkReplicaCount(ctx, tc, scratchRange, 1 /* voterCount */, 2 /* nonVoterCount */)

```

Your next commit fixes this by having `checkReplicaCount` take in a pointer to the range descriptor, I'd recommend moving that to the first commit. 
</blockquote></details>

I see, I've moved this as well as the next fix for preventing the race condition into the same commit.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 646 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MwEhG4W4YiACO5lB3t0:-MwcD3K49TUIIHwwKEi-:bzgcc7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf0ae740dc5b4fff311c16de7adb8429063c2f0/pkg/kv/kvserver/replicate_queue_test.go#L646)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I looked at why the test is failing on CI, and its happening because of the following interaction:

When we mark 2 of these nodes as dead, the `replicateQueue`s of the various leaseholder stores rebalance all those nodes' replicas away. Any of these rebalances may fail if the Raft leader is uncollocated with the leaseholder since that opens the possibility of a race between the Raft snapshot queue on Raft leader racing against the leaseholder's replicateQueue sending a snapshot to the newly added learner during the rebalance.  

This is all fine, except that when this race occurs, the `replicateQueue`'s snapshot fails and we roll the learner back, and mark the new store as "throttled" for 5 seconds (which will make the store not a valid rebalance target for the next 5 seconds). https://github.com/cockroachdb/cockroach/blob/52c2df4fb00d943328dc59df9dc210a40a556ab2/pkg/kv/kvserver/store_pool.go#L44-L53

If the raft snapshot queue wins the race enough times (i.e. around 9 times, which will stall rebalances for 45 seconds), the test fails. 

To work around this, I'd suggest setting that cluster setting to `1ms`. Locally, I couldn't repro the failure when I did that. I'd recommend doing this in the setupFn of both the `Dead` and `Decommissioning` variants of the test. 

```
diff --git a/pkg/kv/kvserver/replicate_queue_test.go b/pkg/kv/kvserver/replicate_queue_test.go
index fdd56450d9..93624a6df5 100644
--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -638,10 +638,16 @@ func TestReplicateQueueDeadNonVoters(t *testing.T) {
 				},
 			},
 		)
+
+		_, err := tc.ServerConn(0).Exec(
+			`SET CLUSTER SETTING server.failed_reservation_timeout='1ms'`,
+		)
+		require.NoError(t, err)
+
 		// Setup a scratch range on a test cluster with 2 non-voters and 1 voter.
 		scratchKey := tc.ScratchRange(t)
 		scratchRange := tc.LookupRangeOrFatal(t, scratchKey)
-		_, err := tc.ServerConn(0).Exec(
+		_, err = tc.ServerConn(0).Exec(
 			`ALTER RANGE DEFAULT CONFIGURE ZONE USING num_replicas = 3, num_voters = 1`,
 		)
 		require.NoError(t, err)

``` 


</blockquote></details>

Wow, thanks for your help digging deep into this one! I see what you mean, I added the fix in for the cluster setting in the first commit, and just waiting for the CI tests to verify the tests now. 


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once you address these last nits. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-:-Mwn9Yy6BrBQfdGkEMn0:b-xy0dvo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @amygao9, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue_test.go, line 646 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MwEhG4W4YiACO5lB3t0:-MwcPur90f8xnXQqQ9H6:b-m499qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf0ae740dc5b4fff311c16de7adb8429063c2f0/pkg/kv/kvserver/replicate_queue_test.go#L646)):*
>but why do the two queues not race when the leader and leaseholder are collocated? 

It's because of the best effort lock on snapshots that place here: https://github.com/cockroachdb/cockroach/blob/d792adf242479ebf764cb2ff5d2055b520898d9a/pkg/kv/kvserver/replica_command.go#L1564

It's best effort because it's in-memory and doesn't prevent this sort of race when the Raft leader and the range leaseholder are not collocated.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 473 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mwn7Ia67ofB_1cTB5fH:-Mwn7Ia77ytxZkDBj6w0:b-6reli7) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L473)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		// check the value of metrics prior to replacement
> ```

nit: let's fix the punctuation.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 813 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mwn90Do7yRAK4oSlkDr:-Mwn90Do7yRAK4oSlkDs:bra79qw) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L813)):*
> ```Go
> 
> // TestReplicateQueueMetrics is an end-to-end test ensuring the replicateQueue
> // voter replica metrics will be updated correctly during upreplication and downreplication.
> ```

nit: wrapping.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 924 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mwn9C477iVfCNBef9uL:-Mwn9C477iVfCNBef9uM:b-28jhqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L924)):*
> ```Go
> 	require.GreaterOrEqualf(
> 		t, currentAddVoterCount, previousAddVoterCount+2,
> 		"expected replica removals to increase by at least 2",
> ```

nit: `s/replica/voter`
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 947 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mwn7U6OC3v0rcpz5sDO:-Mwn7U6OC3v0rcpz5sDP:b-6lieve) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L947)):*
> ```Go
> 	tc *testcluster.TestCluster,
> 	voterMap map[roachpb.NodeID]roachpb.StoreID,
> 	action string,
> ```

nit: using strings in this manner is usually frowned upon, consider just using a `bool` instead.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 6 files at r3, 5 of 5 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @amygao9)

___
*[pkg/kv/kvserver/replicate_queue.go, line 401 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-v93j9_Fd271G5bVA:-Mx-v93k7wun2yThOPIB:bwx1pxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9d856e6a0fd406ce9ce7b535eb1170a9514b75b/pkg/kv/kvserver/replicate_queue.go#L401)):*
> ```Go
> 	// Add replicas.
> 	case AllocatorAddVoter:
> 		return rq.addOrReplaceVoters(ctx, repl, liveVoterReplicas, liveNonVoterReplicas, -1 /* removeIdx */, dryRun)
> ```

Did you mean to remove these inline comments?
___
*[pkg/kv/kvserver/replicate_queue.go, line 250 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-ugxR3cp7LO0qOUmI:-Mx-ugxR3cp7LO0qOUmJ:b-9xrj68) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue.go#L250)):*
> ```Go
> func (metrics *ReplicateQueueMetrics) trackAddReplicaCount(targetType targetReplicaType) {
> 	metrics.AddReplicaCount.Inc(1)
> 	switch targetType {
> ```

For each of these switch statements, we generally prefer including a default case to catch bugs. Something simple like:
```
default:
    panic("unreachable")
```
___
*[pkg/kv/kvserver/replicate_queue.go, line 1326 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-vS6B-TaCWb25OTSQ:-Mx-vS6C1iAMF4vowFf_:b4chnqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue.go#L1326)):*
> ```Go
> 				rq.metrics.trackRebalanceReplicaCount(rebalanceTargetType)
> 			}
> 			if performingSwap && !dryRun {
> ```

nit: consider moving the `if performingSwap {` condition inside the previous block.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 599 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-veVB4Dc32ISaG_tX:-Mx-veVB4Dc32ISaG_tY:bk02xu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L599)):*
> ```Go
> 	}
> 	var store *kvserver.Store
> 	for _, s := range tc.Servers {
> ```

Does the `NodeID` tell us which server to look at? In other words, can we just do `server := tc.Servers[leaseHolder.NodeID-1]` instead of iterating?


<!-- Sent from Reviewable.io -->

amygao9 reviewed (commented):
Addressed all the changes, thanks to both of you for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-:-Mx1-94q9IPvE0uYvRb8:b-ayrxhp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue.go, line 401 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-v93j9_Fd271G5bVA:-Mx0Ocmy3Aclvg81Ee8G:b-kieke3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9d856e6a0fd406ce9ce7b535eb1170a9514b75b/pkg/kv/kvserver/replicate_queue.go#L401)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you mean to remove these inline comments?
</blockquote></details>

Oops, I didn't mean to remove them. Added them back in.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 646 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-MwEhG4W4YiACO5lB3t0:-Mx0H6JmDkzyixPQOf9G:bccp0fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf0ae740dc5b4fff311c16de7adb8429063c2f0/pkg/kv/kvserver/replicate_queue_test.go#L646)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

>but why do the two queues not race when the leader and leaseholder are collocated? 

It's because of the best effort lock on snapshots that place here: https://github.com/cockroachdb/cockroach/blob/d792adf242479ebf764cb2ff5d2055b520898d9a/pkg/kv/kvserver/replica_command.go#L1564

It's best effort because it's in-memory and doesn't prevent this sort of race when the Raft leader and the range leaseholder are not collocated.

</blockquote></details>

Ah, I see. Thanks!
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 599 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx-veVB4Dc32ISaG_tX:-Mx0zT7-8ZGV4uVJYuXn:b-rwcefb) ([raw file](https://github.com/cockroachdb/cockroach/blob/060004029a700b6c1bae894923ff96766f367217/pkg/kv/kvserver/replicate_queue_test.go#L599)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does the `NodeID` tell us which server to look at? In other words, can we just do `server := tc.Servers[leaseHolder.NodeID-1]` instead of iterating?
</blockquote></details>

Yes, you're right! I initially thought that the servers could start in parallel, making the nodeIDs unpredictable. Upon a closer look I realized we don't have the `clusterArgs.ParallelStart` setting on, so the servers should be started in sequential order.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-:-Mx4vqT25KqUz3j1q07w:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @amygao9, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue_test.go, line 851 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vmtA6DqnKUn-Sv8d:-Mx4vmtA6DqnKUn-Sv8e:b-3oak4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L851)):*
> ```Go
> 		tc.(*testcluster.TestCluster),
> 		voterStores,
> 		false,
> ```

`false /* add */`
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 877 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vkIBBfC8ya843JWd:-Mx4vkIBBfC8ya843JWe:b-3oak4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L877)):*
> ```Go
> 		tc.(*testcluster.TestCluster),
> 		voterStores,
> 		false,
> ```

`false /* add */`
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 922 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vh808JvEIjZbtSrc:-Mx4vh808JvEIjZbtSrd:b-433e36) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L922)):*
> ```Go
> 		tc.(*testcluster.TestCluster),
> 		voterMap,
> 		true,
> ```

`true /* add */`
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 963 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vcd-8fFksYkSNZBX:-Mx4vcd-8fFksYkSNZBY:b-5dubud) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L963)):*
> ```Go
> 			}
> 			switch {
> 			case add:
> ```

nit: `if add { ... } else { ... }`


<!-- Sent from Reviewable.io -->

amygao9 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75251)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue_test.go, line 851 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vmtA6DqnKUn-Sv8d:-Mx5HBTe3lmPvb4-qefc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L851)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`false /* add */`
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 877 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vkIBBfC8ya843JWd:-Mx5HAR535cNh3Jgd0nt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L877)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`false /* add */`
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 922 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vh808JvEIjZbtSrc:-Mx5H90o2Ir7nCITwVJW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L922)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`true /* add */`
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue_test.go, line 963 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75251#-Mx4vcd-8fFksYkSNZBX:-Mx5GwBeBqeWJ6dS-crh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aff44252a50f7020f8dfe104ca607dcd2d33a8/pkg/kv/kvserver/replicate_queue_test.go#L963)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `if add { ... } else { ... }`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

amygao9 commented:
TFTRs!

bors r+

amygao9 commented:
bors r+
